src/main/java/fi/aalto/cs/apluscourses/intellij/utils/Interfaces.java

```diff
@@ -88,6 +114,133 @@ public static void refreshPath(VirtualFile file, Runnable callback) {
+     * @param project The current IntelliJ project.
+     * @param courseId ID of the course which contains the exercise in question.
+     * @param exerciseId ID of the exercise.
+     * @param files A map of submittable files, as constructed in {@link SubmitExerciseAction}.
+     * @return True if duplicate check succeeded and the submission should proceed; false if it should be cancelled.
+     */
+    public static boolean checkForDuplicateSubmission(@NotNull Project project,
+                                                      @NotNull String courseId,
+                                                      long exerciseId,
+                                                      @NotNull Map<String, Path> files) {
+      // we want to defer creation of the cache until it's actually necessary
+      if (submissionsCache == null) {
+        Path projectPath = Path.of(Objects.requireNonNull(project.getBasePath()));
+        submissionsCache = new JsonFileCache(
+            projectPath.resolve(Path.of(Project.DIRECTORY_STORE_FOLDER, "a-plus-hashes.json")));
+      }
+
+      try {
+        final MessageDigest shaDigest = MessageDigest.getInstance("SHA-256");
+        final StringBuilder submissionString = new StringBuilder();
+
+        // the iteration order is important so that files always get concatenated in the same order
+        files.entrySet().stream().sorted(Map.Entry.comparingByKey()).forEachOrdered(entry -> {
+          submissionString.append(entry.getKey());
+          submissionString.append('|'); // a separating character that is not in base-64 alphabet
+          submissionString.append(hashFileToString(shaDigest, entry.getValue()));
+          submissionString.append(',');
+        });
+
+        final String cacheKey = "hash_c" + courseId + "_e" + exerciseId;
+        final byte[] finalHashBytes = shaDigest.digest(submissionString.toString().getBytes(StandardCharsets.UTF_8));
+        final String finalHash = Base64.getEncoder().encodeToString(finalHashBytes);
+
+        final JSONObject cachedHashesJson = submissionsCache.getValue(cacheKey, CachePreferences.PERMANENT);
+        final SubmissionHashes cachedHashes = new SubmissionHashes(cachedHashesJson);
+        if (cachedHashesJson == null) {
+          // this is the first hash, so the entry doesn't yet exist in the cache
+          cachedHashes.addHash(finalHash);
+          submissionsCache.putValue(cacheKey, cachedHashes.toJsonObject(), CachePreferences.PERMANENT);
+          return true;
+        }
+
+        if (cachedHashes.containsHash(finalHash)) {
+          if (!DuplicateSubmissionDialog.showDialog()) {
+            return false;
+          }
+        } else {
+          cachedHashes.addHash(finalHash);
+        }
+
+        submissionsCache.putValue(cacheKey, cachedHashes.toJsonObject(), CachePreferences.PERMANENT);
+        return true;
+      } catch (UncheckedIOException | NoSuchAlgorithmException e) {
```

Comment:
Which method may throw `UncheckedIOException` (a RuntimeException). I would narrow down this `try` block, if possible.

Reply:
The method `hashFileToString` can throw IOException. Normally I'd just mark that method with `throws IOException` and handle this instead of `UncheckedIOException`, but unfortunately the invocation of `hashFileToString` is inside a lambda, and for some reason a lambda inside a try-block is not actually covered by that block.

It's essentially a bypass for the IOException being a checked exception; in a sane programming language this shouldn't have been an issue whatsoever ～＿～

Reply:
Oh, I see. What about changing `forEachOrdered` to a normal `for` loop?

Reply:
This would have fixed it! But then the code wouldn't be beautiful, functional, Java 37 -style, but rather imperative, instant legacy code :pp
But one needs to sort the map anyway, so using streams all the way was more natural IMO. I'll change to a for loop if necessary.

Also, I don't really consider that try-block spanning the whole method a bad thing: I wouldn't want a potential exception blocking a user from submitting the exercise.